File: cpp/tests/unit-tests/AccessControlAlgorithmTest.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

#include "libjoynrclustercontroller/access-control/AccessControlAlgorithm.h"
#include "joynr/PrivateCopyAssign.h"
#include "tests/utils/Gtest.h"

#include "joynr/infrastructure/DacTypes/MasterAccessControlEntry.h"
#include "joynr/infrastructure/DacTypes/OwnerAccessControlEntry.h"
#include "joynr/infrastructure/DacTypes/Permission.h"
#include "joynr/infrastructure/DacTypes/TrustLevel.h"

using namespace ::testing;
using namespace joynr;
using namespace joynr::infrastructure::DacTypes;

class AccessControlAlgorithmTest : public ::testing::Test
{
public:
    AccessControlAlgorithmTest() : accessControlAlgorithm()
    {
    }

    ~AccessControlAlgorithmTest() = default;

    void SetUp()
    {
        masterAce = MasterAccessControlEntry(TEST_USER,
                                             TEST_DOMAIN1,
                                             TEST_INTERFACE1,
                                             TrustLevel::LOW,
                                             ALL_TRUST_LEVELS,
                                             TrustLevel::LOW,
                                             ALL_TRUST_LEVELS,
                                             std::string(),
                                             Permission::NO,
                                             ALL_PERMISSIONS);

        mediatorAce = MasterAccessControlEntry(TEST_USER,
                                               TEST_DOMAIN1,
                                               TEST_INTERFACE1,
                                               TrustLevel::LOW,
                                               ALL_TRUST_LEVELS,
                                               TrustLevel::LOW,
                                               ALL_TRUST_LEVELS,
                                               std::string(),
                                               Permission::NO,
                                               ALL_PERMISSIONS);

        ownerAce = OwnerAccessControlEntry(TEST_USER,
                                           TEST_DOMAIN1,
                                           TEST_INTERFACE1,
                                           TrustLevel::LOW,
                                           TrustLevel::LOW,
                                           std::string(),
                                           Permission::NO);
    }

    static const std::vector<TrustLevel::Enum> ALL_TRUST_LEVELS;
    static const std::vector<Permission::Enum> ALL_PERMISSIONS;

protected:
    AccessControlAlgorithm accessControlAlgorithm;
    MasterAccessControlEntry masterAce;
    MasterAccessControlEntry mediatorAce;
    OwnerAccessControlEntry ownerAce;
    static const std::string TEST_USER;
    static const std::string TEST_DOMAIN1;
    static const std::string TEST_INTERFACE1;

private:
    DISALLOW_COPY_AND_ASSIGN(AccessControlAlgorithmTest);
};

const std::string AccessControlAlgorithmTest::TEST_USER("testUser");
const std::string AccessControlAlgorithmTest::TEST_DOMAIN1("domain1");
const std::string AccessControlAlgorithmTest::TEST_INTERFACE1("interface1");
const std::vector<TrustLevel::Enum> AccessControlAlgorithmTest::ALL_TRUST_LEVELS = {
        TrustLevel::LOW, TrustLevel::MID, TrustLevel::HIGH};
const std::vector<Permission::Enum> AccessControlAlgorithmTest::ALL_PERMISSIONS = {
        Permission::NO, Permission::ASK, Permission::YES};

TEST_F(AccessControlAlgorithmTest, permissionWithMasterAceOnly)
{
    masterAce.setDefaultConsumerPermission(Permission::YES);
    masterAce.setDefaultRequiredTrustLevel(TrustLevel::HIGH);
    Permission::Enum consumerPermission = accessControlAlgorithm.getConsumerPermission(
            masterAce,
            boost::optional<MasterAccessControlEntry>(),
            boost::optional<OwnerAccessControlEntry>(),
            TrustLevel::HIGH);

    EXPECT_EQ(Permission::YES, consumerPermission);
}

TEST_F(AccessControlAlgorithmTest, permissionMessageTrustLevelDoesntMatchAce)
{
    masterAce.setDefaultConsumerPermission(Permission::YES);
    masterAce.setDefaultRequiredTrustLevel(TrustLevel::MID);
    Permission::Enum consumerPermission = accessControlAlgorithm.getConsumerPermission(
            masterAce,
            boost::optional<MasterAccessControlEntry>(),
            boost::optional<OwnerAccessControlEntry>(),
            TrustLevel::LOW);

    EXPECT_EQ(Permission::NO, consumerPermission);
}

TEST_F(AccessControlAlgorithmTest, permissionWithAllAceNull)
{

    Permission::Enum consumerPermission = accessControlAlgorithm.getConsumerPermission(
            boost::optional<MasterAccessControlEntry>(),
            boost::optional<MasterAccessControlEntry>(),
            boost::optional<OwnerAccessControlEntry>(),
            TrustLevel::HIGH);

    EXPECT_EQ(Permission::NO, consumerPermission);
}

TEST_F(AccessControlAlgorithmTest, permissionWithMasterAndMediatorAce)
{
    masterAce.setDefaultConsumerPermission(Permission::YES);
    masterAce.setDefaultRequiredTrustLevel(TrustLevel::HIGH);
    masterAce.setPossibleConsumerPermissions(ALL_PERMISSIONS);
    masterAce.setPossibleRequiredTrustLevels(ALL_TRUST_LEVELS);

    mediatorAce.setDefaultConsumerPermission(Permission::ASK);
    mediatorAce.setDefaultRequiredTrustLevel(TrustLevel::LOW);

    Permission::Enum consumerPermission = accessControlAlgorithm.getConsumerPermission(
            masterAce, mediatorAce, boost::optional<OwnerAccessControlEntry>(), TrustLevel::LOW);

    EXPECT_EQ(Permission::ASK, consumerPermission);
}

TEST_F(AccessControlAlgorithmTest, permissionWithMediatorOnly)
{
    mediatorAce.setDefaultConsumerPermission(Permission::YES);
    mediatorAce.setDefaultRequiredTrustLevel(TrustLevel::MID);
    Permission::Enum consumerPermission = accessControlAlgorithm.getConsumerPermission(
            boost::optional<MasterAccessControlEntry>(),
            mediatorAce,
            boost::optional<OwnerAccessControlEntry>(),
            TrustLevel::HIGH);

    EXPECT_EQ(Permission::YES, consumerPermission);
}

TEST_F(AccessControlAlgorithmTest, permissionWithMasterAndInvalidMediatorAce)
{
    std::vector<Permission::Enum> possibleMasterConsumerPermissions;
    possibleMasterConsumerPermissions.push_back(Permission::NO);
    masterAce.setPossibleConsumerPermissions(possibleMasterConsumerPermissions);

    std::vector<Permission::Enum> possibleMediatorConsumerPermissions;
    possibleMediatorConsumerPermissions.push_back(Permission::ASK);
    possibleMediatorConsumerPermissions.push_back(Permission::YES);
    mediatorAce.setPossibleConsumerPermissions(possibleMediatorConsumerPermissions);
    mediatorAce.setDefaultConsumerPermission(Permission::YES);
    mediatorAce.setDefaultRequiredTrustLevel(TrustLevel::MID);

    Permission::Enum consumerPermission = accessControlAlgorithm.getConsumerPermission(
            masterAce, mediatorAce, boost::optional<OwnerAccessControlEntry>(), TrustLevel::HIGH);

    EXPECT_EQ(Permission::NO, consumerPermission);
}

TEST_F(AccessControlAlgorithmTest, permissionWithMasterMediatorAndOwnerAce)
{
    masterAce.setDefaultConsumerPermission(Permission::YES);
    masterAce.setDefaultRequiredTrustLevel(TrustLevel::LOW);

    mediatorAce.setDefaultConsumerPermission(Permission::ASK);
    mediatorAce.setDefaultRequiredTrustLevel(TrustLevel::HIGH);

    ownerAce.setConsumerPermission(Permission::YES);
    ownerAce.setRequiredTrustLevel(TrustLevel::MID);

    Permission::Enum consumerPermission = accessControlAlgorithm.getConsumerPermission(
            masterAce, mediatorAce, ownerAce, TrustLevel::MID);

    EXPECT_EQ(Permission::YES, consumerPermission);
}

TEST_F(AccessControlAlgorithmTest, permissionWithMasterAndOwnerAce)
{
    masterAce.setDefaultConsumerPermission(Permission::ASK);
    masterAce.setDefaultRequiredTrustLevel(TrustLevel::LOW);

    ownerAce.setConsumerPermission(Permission::YES);
    ownerAce.setRequiredTrustLevel(TrustLevel::HIGH);

    Permission::Enum consumerPermission = accessControlAlgorithm.getConsumerPermission(
            masterAce, mediatorAce, ownerAce, TrustLevel::HIGH);

    EXPECT_EQ(Permission::YES, consumerPermission);
}

TEST_F(AccessControlAlgorithmTest, permissionWithOwnerAceOnly)
{
    ownerAce.setConsumerPermission(Permission::YES);
    ownerAce.setRequiredTrustLevel(TrustLevel::HIGH);

    Permission::Enum consumerPermission = accessControlAlgorithm.getConsumerPermission(
            boost::optional<MasterAccessControlEntry>(),
            boost::optional<MasterAccessControlEntry>(),
            ownerAce,
            TrustLevel::HIGH);

    EXPECT_EQ(Permission::YES, consumerPermission);
}

TEST_F(AccessControlAlgorithmTest, permissionWithMediatorAndInvalidOwnerAce)
{
    std::vector<Permission::Enum> possibleMediatorConsumerPermissions;
    possibleMediatorConsumerPermissions.push_back(Permission::NO);
    mediatorAce.setPossibleConsumerPermissions(possibleMediatorConsumerPermissions);

    ownerAce.setConsumerPermission(Permission::ASK);

    Permission::Enum consumerPermission = accessControlAlgorithm.getConsumerPermission(
            boost::optional<MasterAccessControlEntry>(), mediatorAce, ownerAce, TrustLevel::HIGH);

    EXPECT_EQ(Permission::NO, consumerPermission);
}

TEST_F(AccessControlAlgorithmTest, permissionWithMasterAndInvalidOwnerAce)
{
    std::vector<Permission::Enum> possibleMasterConsumerPermissions;
    possibleMasterConsumerPermissions.push_back(Permission::NO);
    masterAce.setPossibleConsumerPermissions(possibleMasterConsumerPermissions);

    ownerAce.setConsumerPermission(Permission::ASK);

    Permission::Enum consumerPermission = accessControlAlgorithm.getConsumerPermission(
            masterAce, boost::optional<MasterAccessControlEntry>(), ownerAce, TrustLevel::HIGH);

    EXPECT_EQ(Permission::NO, consumerPermission);
}

File: java/core/clustercontroller/src/test/java/io/joynr/accesscontrol/AccessControlAlgorithmTest.java
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.accesscontrol;

import java.util.Optional;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import joynr.infrastructure.DacTypes.MasterAccessControlEntry;
import joynr.infrastructure.DacTypes.OwnerAccessControlEntry;
import joynr.infrastructure.DacTypes.MasterRegistrationControlEntry;
import joynr.infrastructure.DacTypes.OwnerRegistrationControlEntry;
import joynr.infrastructure.DacTypes.Permission;
import joynr.infrastructure.DacTypes.TrustLevel;

// Test the access control Algorithm
public class AccessControlAlgorithmTest {

    private static String UID = "testuid";
    private static String DOMAIN = "testdomain";
    private static String INTERFACE = "testinterface";
    private static Permission[] allPermissions = { Permission.NO, Permission.ASK, Permission.YES };
    private static TrustLevel[] allTrustLevels = { TrustLevel.LOW, TrustLevel.MID, TrustLevel.HIGH };

    private AccessControlAlgorithm accessControlAlgorithm;
    private MasterAccessControlEntry masterAce;
    private MasterAccessControlEntry mediatorAce;
    private OwnerAccessControlEntry ownerAce;
    private MasterRegistrationControlEntry masterRce;
    private MasterRegistrationControlEntry mediatorRce;
    private OwnerRegistrationControlEntry ownerRce;

    @Before
    public void setup() {
        this.accessControlAlgorithm = new AccessControlAlgorithm();
        masterAce = new MasterAccessControlEntry(UID,
                                                 DOMAIN,
                                                 INTERFACE,
                                                 TrustLevel.LOW,
                                                 allTrustLevels,
                                                 TrustLevel.LOW,
                                                 allTrustLevels,
                                                 null,
                                                 Permission.NO,
                                                 allPermissions);

        mediatorAce = new MasterAccessControlEntry(UID,
                                                   DOMAIN,
                                                   INTERFACE,
                                                   TrustLevel.LOW,
                                                   allTrustLevels,
                                                   TrustLevel.LOW,
                                                   allTrustLevels,
                                                   null,
                                                   Permission.NO,
                                                   allPermissions);

        ownerAce = new OwnerAccessControlEntry(UID,
                                               DOMAIN,
                                               INTERFACE,
                                               TrustLevel.LOW,
                                               TrustLevel.LOW,
                                               null,
                                               Permission.NO);

        masterRce = new MasterRegistrationControlEntry(UID,
                                                       DOMAIN,
                                                       INTERFACE,
                                                       TrustLevel.LOW,
                                                       allTrustLevels,
                                                       TrustLevel.LOW,
                                                       allTrustLevels,
                                                       Permission.NO,
                                                       allPermissions);

        mediatorRce = new MasterRegistrationControlEntry(UID,
                                                         DOMAIN,
                                                         INTERFACE,
                                                         TrustLevel.LOW,
                                                         allTrustLevels,
                                                         TrustLevel.LOW,
                                                         allTrustLevels,
                                                         Permission.NO,
                                                         allPermissions);

        ownerRce = new OwnerRegistrationControlEntry(UID,
                                                     DOMAIN,
                                                     INTERFACE,
                                                     TrustLevel.LOW,
                                                     TrustLevel.LOW,
                                                     Permission.NO);

    }

    // getConsumerPermission

    @Test
    public void testPermissionWithMasterAceOnly() {
        masterAce.setDefaultConsumerPermission(Permission.YES);
        masterAce.setDefaultRequiredTrustLevel(TrustLevel.HIGH);
        Permission consumerPermission = accessControlAlgorithm.getConsumerPermission(Optional.of(masterAce),
                                                                                     Optional.empty(),
                                                                                     Optional.empty(),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.YES, consumerPermission);
    }

    @Test
    public void testPermissionMessageTrustLevelDoesntMatchAce() {
        masterAce.setDefaultConsumerPermission(Permission.YES);
        masterAce.setDefaultRequiredTrustLevel(TrustLevel.MID);
        Permission consumerPermission = accessControlAlgorithm.getConsumerPermission(Optional.of(masterAce),
                                                                                     Optional.empty(),
                                                                                     Optional.empty(),
                                                                                     TrustLevel.LOW);
        Assert.assertEquals(Permission.NO, consumerPermission);
    }

    @Test
    public void testPermissionWithAllAceNull() {
        Permission providerPermission = accessControlAlgorithm.getConsumerPermission(Optional.empty(),
                                                                                     Optional.empty(),
                                                                                     Optional.empty(),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.NO, providerPermission);
    }

    //------ Mediator overrides master with -----------------------------

    @Test
    public void testPermissionWithMasterAndMediatorAce() {
        masterAce.setDefaultConsumerPermission(Permission.YES);
        masterAce.setDefaultRequiredTrustLevel(TrustLevel.HIGH);
        masterAce.setPossibleConsumerPermissions(allPermissions);
        masterAce.setPossibleRequiredTrustLevels(allTrustLevels);

        mediatorAce.setDefaultConsumerPermission(Permission.ASK);
        mediatorAce.setDefaultRequiredTrustLevel(TrustLevel.LOW);

        Permission providerPermission = accessControlAlgorithm.getConsumerPermission(Optional.of(masterAce),
                                                                                     Optional.of(mediatorAce),
                                                                                     Optional.empty(),
                                                                                     TrustLevel.LOW);
        Assert.assertEquals(Permission.ASK, providerPermission);
    }

    @Test
    public void testPermissionWithMediatorOnly() {
        mediatorAce.setDefaultConsumerPermission(Permission.YES);
        mediatorAce.setDefaultRequiredTrustLevel(TrustLevel.MID);

        Permission providerPermission = accessControlAlgorithm.getConsumerPermission(Optional.empty(),
                                                                                     Optional.of(mediatorAce),
                                                                                     Optional.empty(),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.YES, providerPermission);
    }

    @Test
    public void testPermissionWithMasterAndInvalidMediatorAce() {
        masterAce.setPossibleConsumerPermissions(new Permission[]{ Permission.NO });

        mediatorAce.setPossibleConsumerPermissions(new Permission[]{ Permission.ASK, Permission.YES });
        mediatorAce.setDefaultConsumerPermission(Permission.YES);
        mediatorAce.setDefaultRequiredTrustLevel(TrustLevel.MID);

        Permission consumerPermission = accessControlAlgorithm.getConsumerPermission(Optional.of(masterAce),
                                                                                     Optional.of(mediatorAce),
                                                                                     Optional.empty(),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.NO, consumerPermission);
    }

    //------ Owner overrides master and mediator ---------------------------------

    @Test
    public void testPermissionWithMasterMediatorAndOwnerAce() {
        masterAce.setDefaultConsumerPermission(Permission.YES);
        masterAce.setDefaultRequiredTrustLevel(TrustLevel.LOW);

        mediatorAce.setDefaultConsumerPermission(Permission.ASK);
        mediatorAce.setDefaultRequiredTrustLevel(TrustLevel.HIGH);

        ownerAce.setConsumerPermission(Permission.YES);
        ownerAce.setRequiredTrustLevel(TrustLevel.MID);

        Permission consumerPermission = accessControlAlgorithm.getConsumerPermission(Optional.of(masterAce),
                                                                                     Optional.of(mediatorAce),
                                                                                     Optional.of(ownerAce),
                                                                                     TrustLevel.MID);
        Assert.assertEquals(Permission.YES, consumerPermission);
    }

    @Test
    public void testPermissionWithMasterAndOwnerAce() {
        masterAce.setDefaultConsumerPermission(Permission.ASK);
        masterAce.setDefaultRequiredTrustLevel(TrustLevel.LOW);

        ownerAce.setConsumerPermission(Permission.YES);
        ownerAce.setRequiredTrustLevel(TrustLevel.HIGH);

        Permission consumerPermission = accessControlAlgorithm.getConsumerPermission(Optional.of(masterAce),
                                                                                     Optional.empty(),
                                                                                     Optional.of(ownerAce),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.YES, consumerPermission);
    }

    @Test
    public void testPermissionWithOwnerAceOnly() {
        ownerAce.setConsumerPermission(Permission.YES);
        ownerAce.setRequiredTrustLevel(TrustLevel.HIGH);

        Permission providerPermission = accessControlAlgorithm.getConsumerPermission(Optional.empty(),
                                                                                     Optional.empty(),
                                                                                     Optional.of(ownerAce),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.YES, providerPermission);
    }

    @Test
    public void testPermissionWithMediatorAndInvalidOwnerAce() {
        mediatorAce.setPossibleConsumerPermissions(new Permission[]{ Permission.NO });

        ownerAce.setConsumerPermission(Permission.ASK);

        Permission consumerPermission = accessControlAlgorithm.getConsumerPermission(Optional.empty(),
                                                                                     Optional.of(mediatorAce),
                                                                                     Optional.of(ownerAce),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.NO, consumerPermission);
    }

    @Test
    public void testPermissionWithMasterAndInvalidOwnerAce() {
        masterAce.setPossibleConsumerPermissions(new Permission[]{ Permission.NO });

        ownerAce.setConsumerPermission(Permission.ASK);

        Permission consumerPermission = accessControlAlgorithm.getConsumerPermission(Optional.of(masterAce),
                                                                                     Optional.empty(),
                                                                                     Optional.of(ownerAce),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.NO, consumerPermission);
    }

    // getProviderPermission

    @Test
    public void testPermissionWithMasterRceOnly() {
        masterRce.setDefaultProviderPermission(Permission.YES);
        masterRce.setDefaultRequiredTrustLevel(TrustLevel.HIGH);
        Permission providerPermission = accessControlAlgorithm.getProviderPermission(Optional.of(masterRce),
                                                                                     Optional.empty(),
                                                                                     Optional.empty(),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.YES, providerPermission);
    }

    @Test
    public void testPermissionMessageTrustLevelDoesntMatchRce() {
        masterRce.setDefaultProviderPermission(Permission.YES);
        masterRce.setDefaultRequiredTrustLevel(TrustLevel.MID);
        Permission providerPermission = accessControlAlgorithm.getProviderPermission(Optional.of(masterRce),
                                                                                     Optional.empty(),
                                                                                     Optional.empty(),
                                                                                     TrustLevel.LOW);
        Assert.assertEquals(Permission.NO, providerPermission);
    }

    @Test
    public void testPermissionWithAllRceNull() {
        Permission providerPermission = accessControlAlgorithm.getProviderPermission(Optional.empty(),
                                                                                     Optional.empty(),
                                                                                     Optional.empty(),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.NO, providerPermission);
    }

    //------ Mediator overrides master with -----------------------------

    @Test
    public void testPermissionWithMasterAndMediatorRce() {
        masterRce.setDefaultProviderPermission(Permission.YES);
        masterRce.setDefaultRequiredTrustLevel(TrustLevel.HIGH);
        masterRce.setPossibleProviderPermissions(allPermissions);
        masterRce.setPossibleRequiredTrustLevels(allTrustLevels);

        mediatorRce.setDefaultProviderPermission(Permission.ASK);
        mediatorRce.setDefaultRequiredTrustLevel(TrustLevel.LOW);

        Permission providerPermission = accessControlAlgorithm.getProviderPermission(Optional.of(masterRce),
                                                                                     Optional.of(mediatorRce),
                                                                                     Optional.empty(),
                                                                                     TrustLevel.LOW);
        Assert.assertEquals(Permission.ASK, providerPermission);
    }

    @Test
    public void testPermissionWithMediatorRceOnly() {
        mediatorRce.setDefaultProviderPermission(Permission.YES);
        mediatorRce.setDefaultRequiredTrustLevel(TrustLevel.MID);

        Permission providerPermission = accessControlAlgorithm.getProviderPermission(Optional.empty(),
                                                                                     Optional.of(mediatorRce),
                                                                                     Optional.empty(),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.YES, providerPermission);
    }

    @Test
    public void testPermissionWithMasterAndInvalidMediatorRce() {
        masterRce.setPossibleProviderPermissions(new Permission[]{ Permission.NO });

        mediatorRce.setPossibleProviderPermissions(new Permission[]{ Permission.ASK, Permission.YES });
        mediatorRce.setDefaultProviderPermission(Permission.YES);
        mediatorRce.setDefaultRequiredTrustLevel(TrustLevel.MID);

        Permission providerPermission = accessControlAlgorithm.getProviderPermission(Optional.of(masterRce),
                                                                                     Optional.of(mediatorRce),
                                                                                     Optional.empty(),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.NO, providerPermission);
    }

    //------ Owner overrides master and mediator ---------------------------------

    @Test
    public void testPermissionWithMasterMediatorAndOwnerRce() {
        masterRce.setDefaultProviderPermission(Permission.YES);
        masterRce.setDefaultRequiredTrustLevel(TrustLevel.LOW);

        mediatorRce.setDefaultProviderPermission(Permission.ASK);
        mediatorRce.setDefaultRequiredTrustLevel(TrustLevel.HIGH);

        ownerRce.setProviderPermission(Permission.YES);
        ownerRce.setRequiredTrustLevel(TrustLevel.MID);

        Permission providerPermission = accessControlAlgorithm.getProviderPermission(Optional.of(masterRce),
                                                                                     Optional.of(mediatorRce),
                                                                                     Optional.of(ownerRce),
                                                                                     TrustLevel.MID);
        Assert.assertEquals(Permission.YES, providerPermission);
    }

    @Test
    public void testPermissionWithMasterAndOwnerRce() {
        masterRce.setDefaultProviderPermission(Permission.ASK);
        masterRce.setDefaultRequiredTrustLevel(TrustLevel.LOW);

        ownerRce.setProviderPermission(Permission.YES);
        ownerRce.setRequiredTrustLevel(TrustLevel.HIGH);

        Permission providerPermission = accessControlAlgorithm.getProviderPermission(Optional.of(masterRce),
                                                                                     Optional.empty(),
                                                                                     Optional.of(ownerRce),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.YES, providerPermission);
    }

    @Test
    public void testPermissionWithOwnerRceOnly() {
        ownerRce.setProviderPermission(Permission.YES);
        ownerRce.setRequiredTrustLevel(TrustLevel.HIGH);

        Permission providerPermission = accessControlAlgorithm.getProviderPermission(Optional.empty(),
                                                                                     Optional.empty(),
                                                                                     Optional.of(ownerRce),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.YES, providerPermission);
    }

    @Test
    public void testPermissionWithMediatorAndInvalidOwnerRce() {
        mediatorRce.setPossibleProviderPermissions(new Permission[]{ Permission.NO });

        ownerRce.setProviderPermission(Permission.ASK);

        Permission providerPermission = accessControlAlgorithm.getProviderPermission(Optional.empty(),
                                                                                     Optional.of(mediatorRce),
                                                                                     Optional.of(ownerRce),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.NO, providerPermission);
    }

    @Test
    public void testPermissionWithMasterAndInvalidOwnerRce() {
        masterRce.setPossibleProviderPermissions(new Permission[]{ Permission.NO });

        ownerRce.setProviderPermission(Permission.ASK);

        Permission providerPermission = accessControlAlgorithm.getProviderPermission(Optional.of(masterRce),
                                                                                     Optional.empty(),
                                                                                     Optional.of(ownerRce),
                                                                                     TrustLevel.HIGH);
        Assert.assertEquals(Permission.NO, providerPermission);
    }
}
